src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,829 @@
+})
+
+watch(() => isShowingFilteredCode.value, async function() {
+  // Moves the focus to the code block so that code block-scoped shortcuts still work. That’s necessary because toggling filter mode changes which pre element is rendered. In doing so, the currently focused element is removed from the DOM and in response, the browser moves the focus to document.body.
+  if (document?.activeElement?.tagName === 'PRE') {
+    codeBlock.value?.focus({ preventScroll: true })
+  }
+
+  updateMatchingLineNumbers()
+
+  if (!isShowingFilteredCode.value) {
+    // Waits one tick for the line highlighting to work.
+    await nextTick()
+    // Calls the highlighter function because turning off filter mode causes pre and code element showing highlighted code to be rendered again.
+    highlight()
+  }
+})
+
+onMounted(function() {
+  document.addEventListener('keydown', triggerShortcuts)
+  highlight()
+})
+
+onBeforeUnmount(function() {
+  document.removeEventListener('keydown', triggerShortcuts)
+})
+
+function highlight(): void {
+  if (!(codeBlock.value instanceof HTMLElement)) {
+    return
+  }
+
+  const preElement = codeBlock.value.querySelector('.k-highlighted-code-block')
+  const codeElement = preElement?.querySelector('code')
+
+  if (preElement instanceof HTMLElement && codeElement instanceof HTMLElement) {
+    emit('highlight', {
+      preElement,
+      codeElement,
+      code: props.code,
+      language: props.language,
+      query: query.value,
+      matchingLineNumbers: matchingLineNumbers.value,
+    })
+    updateMatchingLineNumbers()
+  }
+}
+
+function handleSearch(inputValue: string): void {
+  // Ensures that no wasted debouncing takes place when the search input value being emitted is the same as the previous query. This also avoids unnecessarily showing the processing icon.
+  if (inputValue !== query.value) {
+    isProcessingInternally.value = true
+    debouncedHandleSearchInputValue(inputValue)
+  }
+}
+
+function handleSearchInputValue(inputValue: string): void {
+  query.value = inputValue
+  emit('query-change', inputValue)
+  updateMatchingLineNumbers()
+}
+
+function updateMatchingLineNumbers() {
+  isProcessingInternally.value = true
+  regExpError.value = null
+
+  // Avoids searching when one can expect a very large number of results. The numbers here are determined purely by gut feel and not by any scientific reasoning. Feel free to tweak them.
+  const isSmallEnoughForCostlySearch = query.value.length >= 3 || props.code.length < 1000
+  const shouldPerformSearch = query.value.length > 0 && (isRegExpMode.value || (!isRegExpMode.value && isSmallEnoughForCostlySearch))
+  let totalMatchingLineNumbers: number[] = []
+
+  if (shouldPerformSearch) {
+    try {
+      totalMatchingLineNumbers = getMatchingLineNumbers(props.code.toLowerCase(), query.value.toLowerCase(), isRegExpMode.value)
+    } catch (error) {
+      if (error instanceof Error) {
+        regExpError.value = error
+      } else {
+        throw error
```

Comment:
If this isn't an instance of `Error` should we instead do this? (Haven't run the code)

```
throw new Error(error)
```

Reply:
I don’t think so. Presently, this code is not _expected_ to fail with anything but an instance of `Error`. However, should a non-Error throw be introduced, whether accidentally or intentionally, this style of catching ensures that the error isn’t accidentally hidden. The `else` branch here behaves as if we hadn’t caught the error to begin with so I think it’s best to just re-throw whatever was raised.